### PR TITLE
fix: notify operators about how to disable in-cluster on remove

### LINKED ISSUE #14531: Cannot remove local cluster (in-cluster)
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

The local cluster (in-cluster) can't be removed.

```sh
❯ argocd cluster rm in-cluster
Are you sure you want to remove 'in-cluster'? Any Apps deploying to this cluster will go to health status Unknown.[y/n] y
FATA[0001] rpc error: code = NotFound desc = cluster "https://kubernetes.default.svc" not found 
```

I think this might be due to the case that it's a "magic" cluster that is not loaded from a secret?

**To Reproduce**

```sh
argocd cluster rm in-cluster
```

**Expected behavior**

Local cluster is removed from clusters.

**Version**

```shell
❯ argocd version
argocd: v2.7.7+4650bb2.dirty
  BuildDate: 2023-07-05T22:18:46Z
  GitCommit: 4650bb2817c3c81405f40cf77e93ef2b5fb275fb
  GitTreeState: dirty
  GoVersion: go1.20.5
  Compiler: gc
  Platform: darwin/arm64
argocd-server: v2.7.7+4650bb2.dirty
  BuildDate: 2023-07-05T19:42:32Z
  GitCommit: 4650bb2817c3c81405f40cf77e93ef2b5fb275fb
  GitTreeState: dirty
  GoVersion: go1.19.10
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v5.0.1 2023-03-14T01:32:48Z
  Helm Version: v3.11.2+g912ebc1
  Kubectl Version: v0.24.2
  Jsonnet Version: v0.19.1
```

**Logs**

Logs from `argocd-server` pod:

```
time="2023-07-15T20:31:04Z" level=info msg="received unary call /cluster.ClusterService/Delete" grpc.method=Delete grpc.request.content="id:<value:\"https://kubernetes.default.svc\" > " g
rpc.service=cluster.ClusterService grpc.start_time="2023-07-15T20:31:04Z" span.kind=server system=grpc
time="2023-07-15T20:31:04Z" level=info msg="finished unary call with code NotFound" error="rpc error: code = NotFound desc = cluster \"https://kubernetes.default.svc\" not found" grpc.cod
e=NotFound grpc.method=Delete grpc.service=cluster.ClusterService grpc.start_time="2023-07-15T20:31:04Z" grpc.time_ms=9.232 span.kind=server system=grpc
```

### PR DESCRIPTION
**Pull request by morremeyer**
This notifies the operator how they can disable the in-cluster cluster
if they try to remove it.

Closes #14531.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] _not applicable_ ~I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.~ 
* [x] Does this PR require documentation updates? - No.
* [x] _not applicable ~I've updated documentation as required by this PR.~
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] _not applicable_ ~My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.~
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

**Review by gdsoumya — APPROVED** [review]
LGTM

File: cmd/argocd/commands/cluster.go [hunk 1]
```diff
@@ -377,6 +377,13 @@ argocd cluster rm cluster-name`,
 
 			for _, clusterSelector := range args {
 				clusterQuery := getQueryBySelector(clusterSelector)
+
+				// Point users wishing to remove in-cluster to the correct setting
+				if clusterQuery.Server == "https://kubernetes.default.svc" {
```
Comment (gdsoumya) [inline cmd/argocd/commands/cluster.go:382]: There's probably already a constant for the incluster server url we should probably use that

**Review by gdsoumya — CHANGES_REQUESTED** [review]

File: cmd/argocd/commands/cluster.go [hunk 2]
```diff
@@ -377,6 +377,13 @@ argocd cluster rm cluster-name`,
...
+					fmt.Println("The 'in-cluster' cannot be removed. To disable it, set 'cluster.inClusterEnabled: \"false\"' in the argocd-cm ConfigMap.")
+					continue
```
Comment (chengfang) [inline cmd/argocd/commands/cluster.go:384]: shouldn't it exit with code 1, since the rm operation effectively failed?

**Review by todaywasawesome — APPROVED** [review]
LGTM

**Comment by morremeyer** [comment]
Drafting this until I get to fix the tests

**Comment by morremeyer** [comment]
I tried to add a test for this, but failed so far.
I'm afraid I don't understand the tests well enough to write a correct one .

Can anyone help me out with this?

**Comment by morremeyer** [comment]
@todaywasawesome @chengfang @gdsoumya Can you provide guidance for https://github.com/argoproj/argo-cd/pull/16556#issuecomment-1921057553?

**Review by chengfang — APPROVED** [review]

**Comment by morremeyer** [comment]
I looked into how to fix the tests for multiple hours now, and I have no idea how to progress.

Setting up a local environment with make start does not seem to include any of my changes, and setting up a local environment with k3d had the same result.

I couldn’t find any docs on how the custom testing functions work and reading the code didn’t yield any insight for me.

I’ll join the next contributor sync to ask for help on this again, but if that does not result in success, I’ll close the PR.

**Comment by morremeyer** [comment]
I won't get to this in the forseeable future, since the debugging effort is more than I could take time for it currently.

Therefore, I'm closing this. If anybody wants to pick it up, feel free to use the existing commits.

**Pull request closed without merging**